Add note to the guides about action names versus ActionController reserved methods

chrisbloom7 opened the pull request:
### Summary

This commit updates the ActionController Overview section of the Guide to add a note about the potential for method naming conflicts, as detailed below, but stops short of a full list of reserved methods since it's a.) lengthy, and b.) likely to change as internal APIs are updated.

Closes https://github.com/rails/rails/issues/41323

### Other Information

In both Rails 5.2 and Rails 6.1, [defining a controller action method named `config` will result in a `SystemStackError: stack level too deep` exception](https://github.com/rails/rails/issues/41323) for all requests routed to that controller. This is because `ActiveSupport` defines `ActiveSupport::Configurable#config` which is included into `ActionController::Base` by default, and the new config action overrides it. Any actions in the controller will call `render` which eventually will call `logger` which is a configurable attribute which calls `config` which then calls the new `config` action which calls `render` and so on. `config` is [not the only method](https://github.com/rails/rails/issues/41323#issuecomment-860344220) that will trigger this behavior if redefined in a controller: In Rails 6.1, there are 17 methods that would result in `SystemStackError`, 9 that would result in `ArgumentError`, and 3 that would result in a `AbstractController::DoubleRenderError`. Most of these methods are obvious that they should be avoided, like `render`, but some, including `config` since it's never something the user would typically call themselves and its buried deep down in some dependencies, are surprising when encountered and the `SystemStackError` that simply points back to the action method isn't very helpful when trying to debug what has happened.

Commit:
Add note about action names versus ActionController reserved methods

In both Rails 5.2 and Rails 6.1, defining a controller action method
named `config` will result in a `SystemStackError: stack level too deep`
exception for all requests routed to that controller. This is because
`ActiveSupport` defines `ActiveSupport::Configurable#config` which is
included into `ActionController::Base` by default, and the new config
action overrides it. Any actions in the controller will call `render`
which eventually will call `logger` which is a configurable attribute
which calls `config` which then calls the new `config` action which
calls `render` and so on. `config` is not the only method that will
trigger this behavior if redefined in a controller: In Rails 6.1, there
are 17 methods that would result in `SystemStackError` if redefined, 9
that would result in `ArgumentError`, and 3 that would result in a
`AbstractController::DoubleRenderError`. Most of these methods are
obvious that they should be avoided, like `render`, but some, including
`config` since it's never something the user would typically call
themselves and its buried deep down in some dependencies, are surprising
when encountered and the `SystemStackError` that simply points back to
the action method isn't very helpful when trying to debug what has
happened.

This commit updates the ActionController Overview section of the Guide
to add a note about the potential for this conflict, but stops short of
a full list of reserved methods since it's a.) lengthy, and b.) likely
to change as internal APIs are updated.

Closes to https://github.com/rails/rails/issues/41323

zzak commented:
Is the list of known reserved actions documented? 🤔

chrisbloom7 commented:
@zzak I intentionally did not include a list in this PR because it felt too long, and too likely to change as internal APIs change. But if you feel like it would be beneficial to include all or some of them, I can add them. The list of known reserved method names. The ones I've discovered so far are:

| reserved method           | 
|---------------------------| 
| :action_name              | 
| :allow_forgery_protection | 
| :class                    | 
| :config                   | 
| :db_runtime               | 
| :default_form_builder     | 
| :details_for_lookup       | 
| :dispatch                 | 
| :headers                  | 
| :instance_variable_get    | 
| :instance_variables       | 
| :logger                   | 
| :lookup_context           | 
| :main_app                 | 
| :media_type               | 
| :method_for_action        | 
| :process                  | 
| :render                   | 
| :render_to_body           | 
| :rendered_format          | 
| :request                  | 
| :response                 | 
| :response_body            | 
| :run_callbacks            | 
| :send                     | 
| :send_action              | 
| :to_a                     | 
| :view_assigns             | 
| :view_context             | 
| :view_context_class       | 
| :view_runtime             |

chrisbloom7 commented:
(Specific errors that result in overwriting them available [here](https://gist.github.com/chrisbloom7/97a4b10a0135008c68b77016c14b0d93))

zzak commented:
I'm curious if there is an easier way to tell you've tried to use a reserved method then `SystemStackError`.

Outside of that list you created, is there a way to know inside Rails itself? 🙏

chrisbloom7 commented:
@zzak In the initial [issue](https://github.com/rails/rails/issues/41323) there was some discussion about how to detect this at runtime, but it sounded like that wouldn't be possible or worthwhile, mainly because it wouldn't be possible to detect if someone was overriding a method to extend it versus accidentally replacing the original. I'd be interested in hearing alternatives if there's a better way that I'm just not aware of.

lfalcao commented:
@chrisbloom7 Why not have an initializer that in dev and test environment would scan controllers and/or routes for functions with reserved names? Something like `grep -rnw 'app/controllers/' -e 'def config'` - but in Ruby syntax :smile:

chrisbloom7 commented:
@ifalcoa that's a possibility but it would still suffer from the same problem I mentioned above with a JIT reporter about not being able to tell if someone is overwriting a method to extend it versus accidentally replacing the original. In the case of the former, I can imagine it being pretty obnoxious for the framework to continue to pester me about a thing I'm aware of and did intentionally. 

Tagging @rafaelfranca and @gmcgibbon as they had some thoughts on this type of approach. Do either of you think it's worth pursuing that type of solution given the info and suggestions above?

zzak commented:
My thinking is if we can't provide a useful error at run-time, perhaps a linter would be an acceptable place to raise this to the user. Last thing we could do is to add some documentation, including the error for folks searching from the web. The problem with documenting the error is `SystemStackError` is so generic.. 🤔

chrisbloom7 commented:
@zzak I agree with you. `SystemStackError` is very generic and made it hard to track down the root cause when I ran into it. I think a linter opens up a lot more cans of worms though, and to quote a colleague of mine, "the juice might not be worth the squeeze."

For instance, if we have a linter (also applies just as much to a JIT tool) then it needs to be able to distinguish between reopening purposely versus reopening accidentally, so we probably need an AST parser to see if we call `super`. But even that isn't guaranteed to be a perfect signal and there could be reasons for purposely extending it and not calling super or vice versa, so then we need a way to silence those warnings so they don't continue to pester a developer who made that change on purpose. Also, a linter would need to know what methods to examine, and that means we'd need a hardcoded list, and that means we'd need tests to ensure that all the methods in the list throw the errors the linter thinks they will in order to guard against internal API changes that make one or more of those methods no longer problematic to overwrite which could lead to false positives from the linter.

Interestingly, while clicking around through search results for `rails controller SystemStackError` to see what comes up I came across this [4 year old report in `rails/rails`](https://github.com/rails/rails/issues/29217) of the exact same scenario as I reported in my issue, where a comment suggests `So I guess it's not a bug, but it is a bit of a poor experience if you're not aware of the config internals.` I admit this is an anecdotal counter argument, but I've been tinkering with Rails since 1.2.7 and I still tripped over this because I've never needed to pay much attention to Rails' config internals.

Whether we decide a linter is worth the effort or not, it seems to me that a documentation change is a good first step to at least ensuring there's a relevant hit from a trusted source for searches on `rails controller SystemStackError`, and which might cut down on frustration and time to diagnose for the next developer that stumbles into this.

chrisbloom7 commented:
To be clear, I'm not arguing that we don't take that next step of a linter, just calling out some complications that I can foresee. I'm happy to defer to Rails team members on whether it is indeed worthwhile to pursue.

chrisbloom7 commented:
Another thought: if a Rails app has appropriate controller integration tests then it should be easy to tell when something is broken. Perhaps we could augment `ActionDispatch::IntegrationTest` to catch a `SystemStackError` in a test case and offer some guidance there.

zzak commented on guides/source/action_controller_overview.md:
```diff
@@ -65,7 +65,10 @@ The [Layouts and Rendering Guide](layouts_and_rendering.html) explains this in m
 
 Only public methods are callable as actions. It is a best practice to lower the visibility of methods (with `private` or `protected`) which are not intended to be actions, like auxiliary methods or filters.
 
+NOTE: While it's possible to define action methods of any name, some method names are reserved by Action Controller. Accidentally redefining them as actions, or even as auxiliary methods, could result in `SystemStackError` or other exceptions from any route the controller services. Controllers that limit their actions to only the default RESTful [Resource Routing][] actions do not need to worry a ...
```
```suggestion
WARNING: Some method names are reserved by Action Controller. Accidentally redefining them as actions, or even as auxiliary methods, could result in `SystemStackError`. Limiting your controllers to only RESTful [Resource Routing][] actions should not need to worry about this.

NOTE: If you must use a reserved method as an action name, one workaround is to use a custom route to map the reserved method name to your non-reserved action method.
```

zzak replied:
@chrisbloom7 Some wording changes if you could make those and squash your commits I will merge :bow:

chrisbloom7 commented:
@zzak Updated and squashed. I fixed up a grammatical error in the last sentence of the warning:

```diff
- Limiting your controllers to only RESTful [Resource Routing][] actions should not need to worry about this.
+ If you limiting your controllers to only RESTful [Resource Routing][] actions you should not need to worry about this.
```

zzak commented:
Thanks @chrisbloom7! 🎉

chrisbloom7 commented:
Thanks for shepherding this through, @zzak! ❤️
